### PR TITLE
astgen 3.4.0

### PR DESCRIPTION
**Pull request by chenrui333**
Created by `brew bump`

---

Created with `brew bump-formula-pr`.<details>
  <summary>release notes</summary>
  <pre></pre>
</details>

### CODE REVIEW

**Review by ZhongRuoyu — APPROVED** [review]

**Comment by ZhongRuoyu** [comment]
Needs rebase after #138921

File: Formula/astgen.rb [hunk 1]
```diff
@@ -8,7 +8,13 @@ class Astgen < Formula
...
 
   bottle do
-    sha256 cellar: :any_skip_relocation, all: "17e90e5aa92df21c0dd11e8138b86571686459eac7d056b6458c381593df5fe4"
+    sha256 cellar: :any_skip_relocation, arm64_ventura:  "2bba4717d6d280e32524065a452222a0f0c0533041e98eef5d6bd7289a72b15d"
+    sha256 cellar: :any_skip_relocation, arm64_monterey: "2bba4717d6d280e32524065a452222a0f0c0533041e98eef5d6bd7289a72b15d"
+    sha256 cellar: :any_skip_relocation, arm64_big_sur:  "2bba4717d6d280e32524065a452222a0f0c0533041e98eef5d6bd7289a72b15d"
+    sha256 cellar: :any_skip_relocation, ventura:        "2bba4717d6d280e32524065a452222a0f0c0533041e98eef5d6bd7289a72b15d"
+    sha256 cellar: :any_skip_relocation, monterey:       "2bba4717d6d280e32524065a452222a0f0c0533041e98eef5d6bd7289a72b15d"
+    sha256 cellar: :any_skip_relocation, big_sur:        "2bba4717d6d280e32524065a452222a0f0c0533041e98eef5d6bd7289a72b15d"
+    sha256 cellar: :any_skip_relocation, x86_64_linux:   "2bba4717d6d280e32524065a452222a0f0c0533041e98eef5d6bd7289a72b15d"
```
Comment (ZhongRuoyu) [inline Formula/astgen.rb:17]: This looks strange 🤔

**Comment by chenrui333** [comment]
looks like we no longer `all` bottle

image

File: Formula/astgen.rb (hunk 1 above)
Comment (MikeMcQuaid) [reply Formula/astgen.rb]: @ZhongRuoyu not sure what's happening here. @carlocab may have an idea. doubt sharding is to blame as the checksums are the same and the repo is the same. we've only recently fixed the previously broken all bottle situation https://github.com/Homebrew/brew/pull/15819

Comment (carlocab) [reply Formula/astgen.rb]: Can't tell of the top of my head, but I'll have a look.

Comment (carlocab) [reply Formula/astgen.rb]: Generating an `:all` bottle comes from the logic [here](https://github.com/Homebrew/brew/blob/f70a49be2f95dd0e2fd4a7d53550936b1da969f3/Library/Homebrew/dev-cmd/bottle.rb#L636-L639):
```ruby
      all_bottle = !args.no_all_checks? &&
                   (!old_bottle_spec_matches || bottle.rebuild != old_bottle_spec.rebuild) &&
                   tag_hashes.count > 1 &&
                   tag_hashes.uniq { |tag_hash| "#{tag_hash["cellar"]}-#{tag_hash["sha256"]}" }.count == 1
```
Running
```
brew bottle \
  --merge \
  --write \
  --verbose \
  --debug \
  astgen--3.4.0.x86_64_linux.bottle.json \
  astgen--3.4.0.ventura.bottle.json \
  astgen--3.4.0.arm64_ventura.bottle.json \
  astgen--3.4.0.monterey.bottle.json \
  astgen--3.4.0.big_sur.bottle.json \
  astgen--3.4.0.arm64_monterey.bottle.json \
  astgen--3.4.0.arm64_big_sur.bottle.json
```
which is the command that `brew pr-pull` uses to generate the bottle block (via `brew pr-upload`) has `all_bottle` evaluating to `false` for some reason. The bit that is making `all_bottle` evaluate this way is
```ruby
(!old_bottle_spec_matches || bottle.rebuild != old_bottle_spec.rebuild)
```
(the other conditions are `true`).

I'm still working out what this condition is meant to check for, but it seems to be wrong for this formula for some reason.

Comment (carlocab) [reply Formula/astgen.rb]: @Bo98, any ideas? I think you tweaked that particular line a while ago.

**Comment by carlocab** [comment]
Let's merge this now and see if rebottling will help.

**Review by carlocab — APPROVED** [review]

File: Formula/astgen.rb (hunk 1 above)
Comment (carlocab) [reply Formula/astgen.rb]: I suspect this is sharding-related, though. I suggest avoiding the publication of bottles for formulae that will get moved into a subdirectory in a concurrent (but separate) PR. I think it confuses `FormulaVersions`.

Comment (carlocab) [reply Formula/astgen.rb]: > That line is to make sure new dispatched keep-old bottles

Shouldn't it check for `--keep-old` then?

Comment (Bo98) [reply Formula/astgen.rb]: Maybe, but there's also several other things we use `old_bottle_spec` for that's equally broken.

**Comment by carlocab** [comment]
Rebottling seems to have helped. Checksum didn't even change: c940d78bd739c5939ca35ffc713f2ba5d5333c01

**Comment by Bo98** [comment]
Yeah the issue will only happen for the first bottle post-sharding move.

**Pull request merged**